File: FSBaseController/Classes/FSBaseController.m
//
//  FSBaseController.m
//  ShareEconomy
//
//  Created by FudonFuchina on 16/4/13.
//  Copyright © 2016年 FudonFuchina. All rights reserved.
//

#import "FSBaseController.h"

@interface FSBaseController ()

@property (nonatomic, strong) UIView                     *baseLoadingView;
@property (nonatomic, strong) UIView                     *baseBackView;

@property (nonatomic, weak)   NSIndexPath                *cellDeselectIndexPath;
@property (nonatomic, weak)   UITableView                *cellDeselectView;

@end

@implementation FSBaseController {
    BOOL        _onceBase;
    BOOL        _onceBase_viewWillAppear;
    UIView      *_backTapView;
    BOOL        _baseComponentAmounted;
    
    BOOL        _baseAddDidChangeStatusBarOrientationNotification;
}

- (void)dealloc {
#if TARGET_IPHONE_SIMULATOR
    NSLog(@"\n%@ dealloc\n", self.class);
#else
#endif
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)configCellDeselectView:(UITableView *)tableView indexPath:(NSIndexPath *)indexPath {
    _cellDeselectIndexPath = indexPath;
    _cellDeselectView = tableView;
}

static BOOL fitIOS15 = NO;
+ (void)fitIOS15 {
    fitIOS15 = YES;
}

- (void)baseAddBarOrientationChangedNotification {
    _baseAddDidChangeStatusBarOrientationNotification = YES;
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id <UIViewControllerTransitionCoordinator>)coordinator {
    if (_baseAddDidChangeStatusBarOrientationNotification) {
        [self handleOrientationDidChange];
    }
}

- (void)handleOrientationDidChange {
    UIWindowScene *ws = [FSKit currentWindowScene];
    UIInterfaceOrientation fo = ws.interfaceOrientation;
    [self baseHandleChangeStatusBarOrientation: fo];
}

- (void)baseHandleChangeStatusBarOrientation:(UIInterfaceOrientation)orientation {}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (!_onceBase_viewWillAppear) {
        _onceBase_viewWillAppear = YES;
        
        // translucent为YES，self.view布局从屏幕顶部(0,0)开始，如果为NO会从导航栏底部开始
        self.navigationController.navigationBar.translucent = YES;
    }
    
    if (_cellDeselectIndexPath && _cellDeselectView) {
        [_cellDeselectView deselectRowAtIndexPath: _cellDeselectIndexPath animated: YES];
        _cellDeselectIndexPath = nil;
        _cellDeselectView = nil;
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    _isVisibling = YES;
    
//    if (!_onceBase) {
//        _onceBase = YES;
//        
//        self.navigationController.navigationBar.topItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle: self.backBarButtonItemTitle ? : FSKit.appName style: UIBarButtonItemStylePlain target: nil action: nil];
//        self.navigationController.navigationBar.topItem.backBarButtonItem.tintColor = self.view.tintColor;
//    }
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear: animated];
    _isVisibling = NO;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    if (fitIOS15) {
        [self fitIOS15];
    }
    
//    [FSTrack event:NSStringFromClass([self class])];
    
    CGFloat rgb = 0.95;
    self.view.backgroundColor = [UIColor colorWithRed: rgb green: rgb blue: rgb alpha: 1];
    
    _backTapView = [FSViewManager viewWithFrame: self.view.bounds backColor: nil];
    [self.view addSubview: _backTapView];
    
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget: self action: @selector(tapActionBase)];
    [_backTapView addGestureRecognizer: tap];
}

- (void)fitIOS15 {
    if (@available(iOS 15.0, *)) {
        // UINavigationBar
        UINavigationBarAppearance *navigationBarAppearance = [[UINavigationBarAppearance alloc] init];
        navigationBarAppearance.backgroundColor = UIColor.whiteColor;
        self.navigationController.navigationBar.scrollEdgeAppearance = navigationBarAppearance;
        self.navigationController.navigationBar.standardAppearance = navigationBarAppearance;

        // UIToolbar
        UIToolbarAppearance *toolBarAppearance = [[UIToolbarAppearance alloc] init];
        toolBarAppearance.backgroundColor = UIColor.whiteColor;
#if  __IPHONE_OS_VERSION_MAX_ALLOWED >= 150000
        self.navigationController.toolbar.scrollEdgeAppearance = toolBarAppearance;
#endif
        self.navigationController.toolbar.standardAppearance = toolBarAppearance;

        // UITabBar
        UITabBarAppearance *tabBarAppearance = [[UITabBarAppearance alloc] init];
        toolBarAppearance.backgroundColor = UIColor.whiteColor;
#if  __IPHONE_OS_VERSION_MAX_ALLOWED >= 150000
        self.tabBarController.tabBar.scrollEdgeAppearance = tabBarAppearance;
#endif
        self.tabBarController.tabBar.standardAppearance = tabBarAppearance;
    }
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    
    if (!_baseComponentAmounted) {
        _baseComponentAmounted = YES;
        _isVisibling = YES;
        [self componentWillMount];
    }
}

- (void)componentWillMount {
#if TARGET_IPHONE_SIMULATOR
    NSArray *viewControllers = self.navigationController.viewControllers;
    if (viewControllers.count >= 2) {
        UIViewController *from = viewControllers[viewControllers.count - 2];
        NSLog(@"\n%@ FROM\n", from.class);
    }
#else
#endif
    
    if (self.navigationController) {
        self.navigationController.interactivePopGestureRecognizer.enabled = YES;
    }
}

- (void)baseHandleDatas {}
- (void)baseDesignViews {}

- (void)tapActionBase {
    [self.view endEditing:YES];
}

- (FSTapScrollView *)scrollView {
    if (!_scrollView) {
        // 纵坐标即使为0，子视图也是从导航栏下面开始布局
        _scrollView = [[FSTapScrollView alloc] initWithFrame:CGRectMake(0, self.view.safeAreaInsets_fs.top, WIDTHFC, HEIGHTFC - self.view.safeAreaInsets_fs.top)];
        _scrollView.contentSize = CGSizeMake(WIDTHFC, HEIGHTFC + 10);
        _scrollView.showsVerticalScrollIndicator = NO;
//        _scrollView.delaysContentTouches = NO;  // 是干嘛的？
        if (@available(iOS 11.0, *)) {
            _scrollView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
        }
        if (_backTapView) {
            [self.view insertSubview:_scrollView aboveSubview:_backTapView];
        }else{
            [self.view addSubview:_scrollView];
        }
        __weak typeof(self)this = self;
        _scrollView.click = ^(FSTapScrollView *view) {
            [this tapActionBase];
        };
        
        if (self.navigationController) {
            [FSKit fitScrollViewOperate:_scrollView navigationController:self.navigationController];
        }
    }
    return _scrollView;
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

- (void)showWaitView:(BOOL)show {
    if (show) {
        CGFloat blackWidth = 80;
        CGRect blackRect = CGRectMake(WIDTHFC / 2 - 40, HEIGHTFC / 2 - 40, blackWidth, blackWidth);
        if (_baseLoadingView) {
            [self.view bringSubviewToFront:_baseLoadingView];
            _baseLoadingView.frame = [UIScreen mainScreen].bounds;
            _baseBackView.frame = blackRect;
        } else {
            _baseLoadingView = [[UIView alloc] initWithFrame:[UIScreen mainScreen].bounds];
            [self.view insertSubview:_baseLoadingView atIndex:self.view.subviews.count];
            
            _baseBackView = [[UIView alloc] initWithFrame:blackRect];
            _baseBackView.alpha = .7;
            _baseBackView.backgroundColor = [UIColor blackColor];
            _baseBackView.layer.cornerRadius = 6;
            [_baseLoadingView addSubview:_baseBackView];
            
            UIActivityIndicatorView *active = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle: UIActivityIndicatorViewStyleLarge];
            active.frame = CGRectMake(0, 0, _baseBackView.width, _baseBackView.height);
            [active startAnimating];
            [_baseBackView addSubview:active];
        }
    } else {
        [_baseLoadingView removeFromSuperview];
        _baseLoadingView = nil;
    }
}

- (UIView *)fs_bottomView {
    CGFloat h = self.view.safeAreaInsets_fs.bottom + 45;
    FSView *bottomView = [[FSView alloc] initWithFrame:CGRectMake(0, HEIGHTFC - h, WIDTHFC, h)];
    bottomView.backgroundColor = UIColor.whiteColor;
    return bottomView;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // TODO 埋点，记录页面收到内存警告
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: FSBaseController/Classes/FSBaseController.swift
//
//  FSBaseControllerS.swift
//  FSBaseController
//
//  Created by pwrd on 2026/1/29.
//

import Foundation

import FSKit
import FSUIKit

open class FSBaseController: UIViewController {
    
    var             _cellDeselectIndexPath  :   IndexPath?               =       nil
    weak var        _cellDeselectView       :   UITableView?             =       nil
    
    var             _baseAddDidChangeStatusBarOrientationNotification    :   Bool             =       false
    
    var             _onceBase_viewWillAppear:   Bool                     =       false
    var             _isVisibling            :   Bool                     =       false

    var             _back_tap_view          :   UIView?                  =       nil
    
    var             _baseComponentAmounted  :   Bool                     =       false
    
    var             _baseLoadingView        :   UIView?                  =        nil
    var             _baseBackView           :   UIView?                  =        nil

    deinit {
        #if TARGET_IPHONE_SIMULATOR
        print("\(type(of: self)) dealloc")
        #else
        #endif
        
//        let d = FSDate.theLastSecondOfDay(Date())
//        print("FSLog d = \(d)")
        
        let n = Date().timeIntervalSince1970
        if n < 1772207999 {
            print("\(type(of: self)) dealloc")
        }
        
        NotificationCenter.default.removeObserver(self)
    }
    
    open func configCellDeselectView(tableView: UITableView, indexPath: IndexPath) {
        _cellDeselectIndexPath = indexPath
        _cellDeselectView = tableView
    }
    
    static var fitIOS15 : Bool = true
    static public func fitIOS15System() {
        FSBaseController.fitIOS15 = true
    }
    
    open func baseAddBarOrientationChangedNotification() {
        _baseAddDidChangeStatusBarOrientationNotification = true
    }
    
    override open func viewWillTransition(to size: CGSize, with coordinator: any UIViewControllerTransitionCoordinator) {
        if _baseAddDidChangeStatusBarOrientationNotification {
            handleOrientationDidChange()
        }
    }
    
    func handleOrientationDidChange() {
        let ws = FSKit.currentWindowScene()
        guard let ws = ws else {
            return
        }
        
        let fo = ws.interfaceOrientation
        baseHandleChangeStatusBarOrientation(orientation: fo)
    }
    
    open func baseHandleChangeStatusBarOrientation(orientation: UIInterfaceOrientation) {}
    
    open override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if _onceBase_viewWillAppear == false {
            _onceBase_viewWillAppear = true
            
            // translucent为YES，self.view布局从屏幕顶部(0,0)开始，如果为NO会从导航栏底部开始
            self.navigationController?.navigationBar.isTranslucent = true
        }
        
        if (_cellDeselectIndexPath != nil) && (_cellDeselectView != nil) {
            _cellDeselectView!.deselectRow(at: _cellDeselectIndexPath! as IndexPath, animated: true)
            
            _cellDeselectView = nil
            _cellDeselectIndexPath = nil
        }
    }
    
    open override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        _isVisibling = true
    }
        
    open override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        _isVisibling = false
        
        if self.isMovingFromParent {
            self.disappearFraomParent()
        }
    }
    
    open func disappearFraomParent() {}
        
    func fitIOS15() {
        if #available(iOS 15.0, *) {} else {
            return
        }
        
        let navigationBarAppearance = UINavigationBarAppearance()
        navigationBarAppearance.backgroundColor = UIColor.white
        self.navigationController?.navigationBar.scrollEdgeAppearance = navigationBarAppearance
        self.navigationController?.navigationBar.standardAppearance = navigationBarAppearance
        
        let toolBarAppearance = UIToolbarAppearance()
        toolBarAppearance.backgroundColor = UIColor.white
        self.navigationController?.toolbar.scrollEdgeAppearance = toolBarAppearance
        self.navigationController?.toolbar.standardAppearance = toolBarAppearance
        
        
        let tabBarAppearance = UITabBarAppearance()
        toolBarAppearance.backgroundColor = UIColor.white
        self.tabBarController?.tabBar.scrollEdgeAppearance = tabBarAppearance
        self.tabBarController?.tabBar.standardAppearance = tabBarAppearance
    }
    
    open override func viewDidLoad() {
        super.viewDidLoad()
        
        if FSBaseController.fitIOS15 {
            fitIOS15()
        }
                
        let rgb = 0.95
        self.view.backgroundColor = UIColor.init(red: rgb, green: rgb, blue: rgb, alpha: 1.0)
                
        _back_tap_view = UIView(frame: self.view.bounds)
        self.view.addSubview(_back_tap_view!)
                
        let tap = UITapGestureRecognizer.init(target: self, action: #selector(tapActionBase))
        _back_tap_view?.addGestureRecognizer(tap)
    }
    
    @objc open func tapActionBase() {
        self.view.endEditing(true)
    }
    
    open override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        if _baseComponentAmounted == false {
            _baseComponentAmounted = true
            
            _isVisibling = true
            
            componentWillMount()
        }
    }
    
    open func componentWillMount() {
        
        #if TARGET_IPHONE_SIMULATOR
        let vcs = self.navigationController?.viewControllers
        if vcs?.count >= 2 {
            let from = vcs?.index(of: vcs?.count - 2)
            print("\n \(type(of: from)) From \n")
        }
        #else
        #endif
        
        if self.navigationController != nil {
            self.navigationController?.interactivePopGestureRecognizer?.isEnabled = true
        }
    }
    
    open func baseHandleDatas() {}
    open func baseDesignViews() {}
    
    open lazy var scrollView: FSTapScrollView = {
        
        let scrollView = FSTapScrollView(frame: CGRect(x: 0, y: self.view.safeAreaInsets.top, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height - self.view.safeAreaInsets.top))
        scrollView.contentSize = CGSizeMake(UIScreen.main.bounds.size.width, UIScreen.main.bounds.size.height + 10)
        scrollView.showsVerticalScrollIndicator = false
        scrollView.contentInsetAdjustmentBehavior = .never
        scrollView.click = { [weak self] view in
            self?.tapActionBase()
        }

        if _back_tap_view == nil {
            self.view.addSubview(scrollView)
        } else {
            self.view.insertSubview(scrollView, aboveSubview: _back_tap_view!)
        }
        
//        let tap = UITapGestureRecognizer.init(target: self, action: #selector(tapActionBase))
//        scrollView.addGestureRecognizer(tap)
        
        if self.navigationController != nil {
            FSKit.fitScrollViewOperate(scrollView, navigationController: self.navigationController)
        }
        
        return scrollView
    }()

    open override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
    
    open func showWaitView(_ show: Bool) {
        
        if show {
            let blackWidth = 80
            let blackRect = CGRect(x: Int(UIScreen.main.bounds.size.width / 2 - 40.0), y: (Int(UIScreen.main.bounds.size.height) / 2 - 40), width: blackWidth, height: blackWidth)
            
            if _baseLoadingView != nil {
                self.view.bringSubviewToFront(_baseLoadingView!)
                _baseLoadingView!.frame = UIScreen.main.bounds
                _baseBackView?.frame = blackRect
            } else {
                _baseLoadingView = UIView(frame: UIScreen.main.bounds)
                self.view.addSubview(_baseLoadingView!)
                
                _baseBackView = UIView(frame: blackRect)
                _baseBackView?.alpha = 0.7
                _baseBackView?.backgroundColor = UIColor.black
                _baseBackView?.layer.cornerRadius = 6
                _baseLoadingView?.addSubview(_baseBackView!)
                
                let active = UIActivityIndicatorView.init(style: .large)
                active.frame = CGRect(x: 0, y: 0, width: Int(_baseBackView!.frame.size.width), height: Int(_baseBackView!.frame.size.height))
                active.startAnimating()
                _baseBackView?.addSubview(active)
                
            }
            
        } else {
            _baseLoadingView?.removeFromSuperview()
            _baseLoadingView = nil
        }
    }
    
    open func fs_bottomView() -> UIView {
        let h = self.view.safeAreaInsets.bottom + 45
        let bottomView = UIView(frame: CGRect(x: 0, y: UIScreen.main.bounds.size.height - h, width: UIScreen.main.bounds.size.width, height: h))
        bottomView.backgroundColor = UIColor.white
        return bottomView
    }
    
}
